### ISSUE TITLE
Path traversal vulnerability when serving static file

### ISSUE THREAD

**Issue by splitline** [GenieFramework/Genie.jl#495]
I notice that there is a path traversal vulnerability in Genie.jl.

**To reproduce**

Here is a pretty simple example:
```julia
using Genie, Genie.Router
route("/") do
  "Hello World"
end
up(8888)
```
Yep, it's just a simple hello world application.

Then if we perform a request to `http://127.0.0.1:8888////etc/hosts`, it should show the content of your `/etc/hosts` file:

<img width="922" alt="圖片" src="https://user-images.githubusercontent.com/10939796/156415970-699b0338-6be8-4178-8a0c-24284e1e0c84.png">

And if there contains a `./public` directory in the work directory of Genie.jl application, request to `http://127.0.0.1:8888/../../../../../../../../../etc/hosts` should also work.

<img width="872" alt="圖片" src="https://user-images.githubusercontent.com/10939796/156419684-589de6cb-0b43-403f-b5c8-919d8e29e7de.png">

**How it happened**

When the server receives a request, it'll handle by `route_request` function and check if the incoming request is a static file:
https://github.com/GenieFramework/Genie.jl/blob/09d875b421deccc618c9dc332bbefb4d564447ed/src/Router.jl#L137-L141

Then let's take a look at `is_static_file`:
https://github.com/GenieFramework/Genie.jl/blob/09d875b421deccc618c9dc332bbefb4d564447ed/src/Router.jl#L1039-L1041

First it converts the request path using `HTTP.URIs.URI(resource).path`. During this process path like `////etc/hosts` will become `//etc/hosts`, since the first two slashes will be recognized as a part of the URI scheme.

Then the result (`//etc/hosts`) will be passed into `file_path` function:

https://github.com/GenieFramework/Genie.jl/blob/09d875b421deccc618c9dc332bbefb4d564447ed/src/Router.jl#L1153-L1156

Here is the main cause of this vulnerability. 

Let's see a quick example: `joinpath("any_path", "/etc/hosts")` will return `/etc/hosts`. Long story short, if one argument starts with slash, then it'll become the root path instead of concat after the previous argument.

So if we pass `//etc/hosts` into `file_path`, then it strips the first slash, but there still remains one slash after the stripped slash, so the result of `file_path` will still be `/etc/hosts`.

And yes, that is a file, so it goes to `serve_static_file`, then the vulnerability happens.

By the way, for the `../` one, no matter you trimmed the leading slash or not, it always happens.

**How to prevent this**

1. Trim all the leading slash
2. Check if there is any relative path in the request path. If there is one maybe we should resolve it first before serve it as static file, or just simply return a "400 bad request"

**Comment by splitline** [comment]
By the way, maybe we should create a [security policy](https://github.com/GenieFramework/Genie.jl/security/policy) for reporting this kind of security issues?

**Comment by AbhimanyuAryan** [comment]
thanks a ton @splitline I'll look into this as well. Thanks for using Genie and digging for possible vulnerabilities. You are awesome 😎

**Comment by AbhimanyuAryan** [comment]
> By the way, maybe we should create a [security policy](https://github.com/GenieFramework/Genie.jl/security/policy) for reporting this kind of security issues?

yes already have this in TODO. Will add

**Comment by essenciary** [comment]
@splitline Thank you for reporting this and including all the details, very helpful and much appreciated. 
This has now been patched in v4.16 (currently being registered) -- tests included.

**Comment by essenciary** [comment]
I will close this and then delete it - after which I will announce to our users to upgrade to v4.16. 
Thanks again!
